api/inc/rpc_gateway.h

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright (c) 2016, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef __UVISOR_API_RPC_GATEWAY_H__
+#define __UVISOR_API_RPC_GATEWAY_H__
+
+#include "api/inc/rpc_gateway_exports.h"
+#include "api/inc/rpc.h"
+#include "api/inc/uvisor_exports.h"
+#include <stdint.h>
+
+/* ldr pc, [pc, #<label - instr + 4>]
+ * LDR (immediate) - ARMv7M ARM section A7.7.42
+ * 1111;1 00 0; 0 10 1; <Rn - 1111>; <Rt - 1111>; <imm12> (Encoding T3) */
+#define LDR_PC_PC_IMM_OPCODE(instr, label) \
+    ((uint32_t) (0xF000F8DFU | ((((uint32_t) (label) - ((uint32_t) (instr) + 4U)) & 0xFFFU) << 16U)))
+
+/** Synchronous RPC Gateway
+ *
+ * This macro declares a new function pointer (with no name mangling) named
+ * `gw_name` to perform a remote procedure call (RPC) to the target function
+ * given by `fn_name`. RPCs are assembled into a read-only flash structure that
+ * is read and validated by uVisor before performing the operation.
+ *
+ * Create function with following signature:
+ * UVISOR_EXTERN fn_type fn_name(uint32_t a, uint32_t b);
+ *
+ * @param box_name[in] The name of the source box as declared in
+ *                     `UVISOR_BOX_CONFIG`
+ * @param gw_name[in]  The new, callable function pointer for performing RPC
+ * @param fn_name[in]  The function being designated as an RPC target
+ * @param fn_type[in]  The return type of the function being designated as an
+ *                     RPC target
+ * @param __VA_ARGS__  The type of each parameter passed to the target
+ *                     function. There can be up to 4 parameters. Each
+ *                     parameter must be no more than uint32_t in size.
+ */
+#define UVISOR_BOX_RPC_GATEWAY_SYNC(box_name, gw_name, fn_name, fn_type, ...) \
+    UVISOR_STATIC_ASSERT(sizeof(fn_type) <= sizeof(uint32_t), gw_name ## _return_type_too_big); \
+    _UVISOR_BOX_RPC_GATEWAY_ARG_CHECK(gw_name, __VA_ARGS__) \
```

Comment:
I'm quite familiar with ##__VA_ARGS__. I try to avoid GCC-specific extensions where possible in user/library code. (GCC-specific stuff is fine in uVisor itself.)

For the RPC creating gateway macros, we can get by without GCC-specific extensions as we do require 1 argument. If the function you are writing a gateway to doesn't take any parameters, specify `void`.

I'll add a comment in the doxygen notes that at least one parameter is required.